### ISSUE TITLE
Warn about functions that do the same thing.

### ISSUE THREAD

**Issue by oilel** [rust-lang/rust-clippy#10512]
### What it does

Warn about functions that do the same thing.

### Example

```
use rand::*;
fn a() -> u8 {
    let mut rng = thread_rng();
    rng.gen()
}
fn b() -> u8 {
    let mut rng = thread_rng();
    rng.gen()
}
fn main(){
    println!("{}\n{}", a(), b())
}
```

Both a() and b() generate a random u8, but clippy doesn't warn about they're the same.

**Comment by oilel** [comment]
It can become more complex:
```
use rand::*;
fn a() -> u8 {
    let mut rng = thread_rng();
    rng.gen()
}
fn b() -> u8 {
    let mut rng = thread_rng();
    let useless = 76;
    rng.gen()
}
fn main(){
    println!("{}\n{}", a(), b())
}
```
I add a useless initialization, so the functions are still the same, clippy should still warn on it even if I added a useless initialization.

**Comment by schubart** [comment]
About the useless initialisation: I think the new lint you're proposing doesn't need to handle this. Other, existing lints should warn that `useless` is useless. Then, after the user fixes that, your proposed new lint can kick in and detect that `a()` and `b()` are now equal.

**Comment by oilel** [comment]
> I think the new lint you're proposing doesn't need to handle this.

If a function is defined as to return 2u8, another function is defined as to return (1+1) as u8, does it need to handle that? Since 1+1=2, both are u8, they're the same.

**Comment by schubart** [comment]
I was only referring to your comment "It can become more complex" where you mention a useless initialisation. I was not trying to say that the proposed lint isn't useful.

However, in general, it is not possible to decide if two functions are equivalent. See for example https://stackoverflow.com/q/1132051 for an explanation.

Such a lint might still be useful: It could flag simple cases were two functions *are* equivalent.
